### PR TITLE
4.x: Allowed values are now checked when validating a builder.

### LINKED ISSUE #6231: `AnnotationAndValue` is now limited to values that are a single string - no support for arrays, nested annotations (this may be required as we implement Helidon features)

### LINKED ISSUE #6403: ConfiguredOption allowedValues needs support in Builder
The generated Validators should support `ConfiguredOption.allowedValues`, currently this is not supported.

### PR DESCRIPTION
**Pull request by tomas-langer**
### Description
When using `@ConfiguredOption(allowedValues=...)` the generated builders now validate (during build) that the option is within the allowed values.

- for an option, if not `null`, a check is made of its `String.valueOf()` against the String allowed value set
- for a list of options, each value in the list is check (if any)

This change required modifying `Annotation` to support typed values and nested annotations, this has been done as well.

### Documentation
This enhancement enforces an existing API in `ConfiguredOption`.
Annotation now supports nested annotations and typed values. Methods to obtain a typed value are now added to the `Annotation` type (in `helidon-common-types`)

Resolves #6403 
Resolves #6231

### CODE REVIEW

File: builder/processor/src/main/java/io/helidon/builder/processor/PrototypeProperty.java [hunk 1]
```diff
@@ -115,6 +116,10 @@ static PrototypeProperty create(ProcessingContext processingContext,
         );
     }
 
+    public Field.Builder fieldDeclaration(boolean isBuilder) {
```
Comment (Verdent) [inline builder/processor/src/main/java/io/helidon/builder/processor/PrototypeProperty.java:119]: not needed `public`, this is package private record.

File: common/processor/processor/src/main/java/io/helidon/common/processor/ToAnnotationValueVisitor.java [hunk 2]
```diff
@@ -21,93 +21,97 @@
 
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.AnnotationValue;
-import javax.lang.model.element.AnnotationValueVisitor;
 import javax.lang.model.element.VariableElement;
 import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
 
-class ToStringAnnotationValueVisitor implements AnnotationValueVisitor<String, Object> {
+class ToAnnotationValueVisitor implements javax.lang.model.element.AnnotationValueVisitor<Object, Object> {
```
Comment (Verdent) [inline common/processor/processor/src/main/java/io/helidon/common/processor/ToAnnotationValueVisitor.java:28]: Is this fully qualified package correct here? I mean, what is the reason for this change?

File: common/types/src/main/java/io/helidon/common/types/Annotation.java [hunk 3]
```diff
@@ -40,8 +67,8 @@ public interface Annotation extends AnnotationBlueprint, Prototype.Api, Comparab
      *
      * @return a new builder
      */
-    static Annotation.Builder builder() {
-        return new Annotation.Builder();
+    static Builder builder() {
```
Comment (Verdent) [inline common/types/src/main/java/io/helidon/common/types/Annotation.java:70]: is this what IDE did? Since I think for merging purposes it would be better to leave it with `Annotation` prefix.

File: common/types/src/main/java/io/helidon/common/types/Annotation.java [hunk 4]
```diff
@@ -374,7 +401,7 @@ public int hashCode() {
     /**
      * Fluent API builder for {@link Annotation}.
      */
-    class Builder extends Annotation.BuilderBase<Annotation.Builder, Annotation> implements io.helidon.common.Builder<Annotation.Builder, Annotation> {
+    class Builder extends BuilderBase<Builder, Annotation> implements io.helidon.common.Builder<Builder, Annotation> {
```
Comment (Verdent) [inline common/types/src/main/java/io/helidon/common/types/Annotation.java:404]: The same as above

**Review by Verdent — CHANGES_REQUESTED** [review]

File: builder/tests/builder/src/main/java/io/helidon/builder/test/testsubjects/MyConfigBeanBlueprint.java [hunk 5]
```diff
@@ -52,7 +52,7 @@ interface MyConfigBeanBlueprint {
      */
     @ConfiguredOption(key = "port", value = "8080", allowedValues = {
             @ConfiguredValue(value = "8080", description = "t1"),
-            @ConfiguredValue(value = "false", description = "t2")})
+            @ConfiguredValue(value = "80", description = "t2")})
```
Comment (trentjeff) [inline builder/tests/builder/src/main/java/io/helidon/builder/test/testsubjects/MyConfigBeanBlueprint.java:55]: what is really needed here is a range of good values.  I think ultimately for this we need to refer to a method or class to instantiate to perform dynamic validations for complex cases.

File: integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java [hunk 6]
```diff
@@ -114,28 +115,27 @@ void potentialAuthStrategies() {
         assertThat(cfg.potentialAuthStrategies(),
                    contains("config", "config-file"));
 
-        config = createTestConfig(ociAuthConfigStrategies("", ""))
+        // this must fail, as empty string is not an allowed strategy
```
Comment (trentjeff) [inline integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java:118]: ok, fine, but please ensure there is a test that when everything is cleared the `potentialAuthStrategies` remain as before.

**Review by trentjeff — CHANGES_REQUESTED** [review]

File: common/processor/processor/src/main/java/io/helidon/common/processor/ToAnnotationValueVisitor.java (hunk 2 above)
Comment (tomas-langer) [reply common/processor/processor/src/main/java/io/helidon/common/processor/ToAnnotationValueVisitor.java:28]: Fixed, I renamed it to `AnnotationValueVisitor` which was incorrect, as it was in conflict with the interface name. Renamed, forgot to fix the FQ name. Original name was no longer correct

File: common/types/src/main/java/io/helidon/common/types/Annotation.java (hunk 3 above)
Comment (tomas-langer) [reply common/types/src/main/java/io/helidon/common/types/Annotation.java:70]: yes, will restore original approach

File: builder/tests/builder/src/main/java/io/helidon/builder/test/testsubjects/MyConfigBeanBlueprint.java (hunk 5 above)
Comment (tomas-langer) [reply builder/tests/builder/src/main/java/io/helidon/builder/test/testsubjects/MyConfigBeanBlueprint.java:55]: `false` is never a good value for an integer port. Not sure what you mean by this. If you need better validation, you can always have a custom decorator

File: integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java (hunk 6 above)
Comment (tomas-langer) [reply integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java:118]: Empty string is not clearing anything. If you think it does, then the test was very wrong in the first place. Empty string is a valid value.
I will check if I can do what you are asking.

File: builder/processor/src/main/java/io/helidon/builder/processor/GenerateAbstractBuilder.java [hunk 7]
```diff
@@ -516,7 +533,9 @@ private static void validatePrototypeMethod(InnerClass.Builder classBuilder, Typ
                 .superPrototype()
                 .ifPresent(it -> validateBuilder.addLine("super.validatePrototype();"));
 
-        if (typeContext.propertyData().hasRequired() || typeContext.propertyData().hasNonNulls()) {
+        if (typeContext.propertyData().hasRequired()
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply builder/processor/src/main/java/io/helidon/builder/processor/GenerateAbstractBuilder.java:536]: Fixed

File: builder/processor/src/main/java/io/helidon/builder/processor/PrototypeProperty.java (hunk 1 above)
Comment (tomas-langer) [reply builder/processor/src/main/java/io/helidon/builder/processor/PrototypeProperty.java:119]: Fixed

File: integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java (hunk 6 above)
Comment (tomas-langer) [reply integrations/oci/sdk/runtime/src/test/java/io/helidon/integrations/oci/sdk/runtime/OciExtensionTest.java:118]: The test you ask for already exists - you have a test that sends `null`, so it validates what happens if these options are not configured.

**Review by Verdent — APPROVED** [review]

**Review by trentjeff — APPROVED** [review]
